### DISCUSSION TITLE
Best Practices for Customizing Services

### DISCUSSION THREAD

**Discussion by bmcandr** [developmentseed/eoAPI#161] [General]
Hey all!

I am a current user (and huge fan) of several of the base services included in `eoAPI`. I have built an app on top of `stac-fastapi-pgstac` and added custom functionality to a fork of `titiler-pgstac` to meet some of our specific business needs. The emergent features enabled by the interconnections between services in `eoAPI` are compelling and I am interested in potentially adopting this stack, but it is unclear (to me, at least) what the best practices are for adding custom functionality to the base services. Or, is `eoAPI` meant for organizations that just need an easy way to deploy this stack as-is?

**Comment by vincentsarago** [comment]
we do need to update https://eoapi.dev/customization/ for clarity! 

The eoapi project is split into multiple things:
- [eoapi-cdk](https://github.com/developmentseed/eoapi-cdk): helps deploy all the eoapi services on AWS using serverless functions
- [eoapi-k8s ](https://github.com/developmentseed/eoapi-k8s): helps deploy all the eoapi services with Kubernetes
- the eoAPI repo (here) represent a custom version of `eoapi` where we experiment with runtimes and UI (and it's set to use eoapi-cdk) for deployment

https://github.com/developmentseed/eoAPI/tree/main/runtime/eoapi/raster is an example of how to customize titiler-pgstac and https://github.com/developmentseed/eoAPI/blob/main/infrastructure/aws/cdk/app.py#L137-L187 is the CDK code used for deployment 

Easier ways for customization will be provided with https://github.com/developmentseed/eoapi-template (at least if you are using eoapi-cdk)

**Reply by bmcandr** [reply]
Thanks!

>the eoAPI repo (here) represent a custom version of eoapi where we experiment with runtimes and UI

This is helpful context. It wasn't clear from the documentation that this repo is more demonstrative and contains experimental features. If I wanted to take advantage of these features and interconnections _and_ add custom functionality to the individual service runtimes, would you recommend forking this repo and modifying each service as necessary (as shown in `runtime/eoapi/raster` and hinted at a bit [here](https://github.com/developmentseed/eoAPI#deployment-with-custom-runtimes))? And adapt `eoapi-k8s`, which looks to be a WIP, to fit our deployment environment (Azure K8s Service)?

Apologies if I'm not groking this--I'm admittedly out of my depth on the IaC/deployment parts so I'm trying to focus on understanding how we can leverage (and stay up to date with) the cool things you're doing in this repo and build on top of it. I'll have to get help from our DevOps engineer to sort out the deployment!
